subprojects/core/src/main/java/org/gradle/api/reflect/TypeOf.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.reflect;
+
+import com.google.common.base.Objects;
+import com.google.common.reflect.TypeToken;
```

Comment:
We would use it to compute its Kotlin type string representation as part of generating Kotlin source code defining extension properties to provide convenient access to the registered project extensions (obtained from the project extensions schema). 

Here's a working Kotlin example with the kind of thing we want to do on gradle-script-kotlin:

```kotlin
package samples

import java.lang.reflect.*

/**
 * This is a example of the type of function we want to be able to write
 * on the gradle-script-kotlin side.
 */
fun kotlinTypeNameFor(type: Type): String =
    when (type) {
        is ParameterizedType -> {
            (kotlinTypeNameFor(type.rawType)
                + type.actualTypeArguments.joinToString(
                      prefix = "<", postfix = ">", transform = ::kotlinTypeNameFor))
        }
        is GenericArrayType -> {
            "Array<${kotlinTypeNameFor(type.genericComponentType)}>"
        }
        is WildcardType -> {
            kotlinTypeNameFor(type.upperBounds[0])
        }
        is Class<*> -> {
            type.simpleName
        }
        else -> {
            throw NotImplementedError()
        }
    }

fun printExtensions(schema: Map<String, Type>) {
    schema.forEach { (name, type) ->
        val typeString = kotlinTypeNameFor(type)
        println("val Project.$name get() = extensions.getByName(\"$name\") as $typeString")
    }
}

interface TypeOf<T : Any>

inline fun <reified T : Any> typeOf(): Type =
    ((object : TypeOf<T> {}).javaClass.genericInterfaces[0] as ParameterizedType).actualTypeArguments[0]

// Prints
// val Project.myExtension get() = extensions.getByName("myExtension") as Array<List<String>>
fun main(args: Array<String>) {
    val schema = mapOf("myExtension" to typeOf<Array<List<String>>>())
    printExtensions(schema)
}
```